server/storage/fs_localCache.go

```diff
@@ -0,0 +1,151 @@
+		NumCounters: 1e7,     // number of keys to track frequency of (10M).
+		MaxCost:     1 << 30, // maximum cost of cache (1GB).
+		BufferItems: 64,      // number of keys per Get buffer.
+		OnEvict: func(item *ristretto.Item) {
+			cached := item.Value.(*localCachedFSValue)
+			log.Debugf("localCache: OnEvict %s", cached.name)
+			remove(cached.name)
+		},
+		OnReject: func(item *ristretto.Item) {
+			cached := item.Value.(*localCachedFSValue)
+			log.Debugf("localCache: OnReject %s", cached.name)
+			remove(cached.name)
+		},
+		Metrics: isDev,
+		/**
+		 * Although tempting to use, this is called when items are replaced,
+		 * which would result in unnecessary removals from the backing fs.
+		 */
+		// OnExit: func(value interface{}) {
+		// 	cached := value.(*LocalCachedFSValue)
+		// 	log.Debugf("localCache: OnExit %s", cached.name)
+		// 	os.Remove(path.Join(root, cached.name))
+		// },
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Hydrate the cache on Open
+	log.Debugf("localCache: hydrating...")
+	filepath.WalkDir(root,
+		func(path string, entry os.DirEntry, err error) error {
+			if err != nil {
+				return err
+			}
+			info, err := entry.Info()
+			if err != nil {
+				return err
+			}
+			name, err := filepath.Rel(root, path)
+			if err != nil {
+				return err
+			}
+			cost := info.Size()
+			modtime := info.ModTime()
+			if !cache.SetWithTTL(name, &localCachedFSValue{name: name, modtime: modtime}, cost, TTL) {
+				remove(name)
+			}
+			return nil
+		})
+
+	cache.Wait()
+
+	if isDev {
+		log.Debugf("localCache: hydration metrics %s", cache.Metrics.String())
+		cache.Metrics.Clear()
+	}
+
+	return &localCacheFSLayer{
+		log:     log,
+		backing: backing,
+		cache:   cache,
+		remove:  remove,
+		TTL:     TTL,
+	}, nil
+}
+
+type localCacheFSLayer struct {
+	log     *logx.Logger
+	backing FSConn
+	cache   *ristretto.Cache
+	remove  func(string)
+	TTL     time.Duration
+}
+
+type localCachedFSValue struct {
+	name    string
+	modtime time.Time
+}
+
+func (fs *localCacheFSLayer) Exists(name string) (found bool, modtime time.Time, err error) {
+	fs.cache.Wait()
+	value, found := fs.cache.Get(name)
+	if found {
+		cached := value.(*localCachedFSValue)
+		modtime = cached.modtime
+	}
+	return
+}
+
+func (fs *localCacheFSLayer) ReadFile(name string) (file io.ReadSeekCloser, err error) {
+	value, found := fs.cache.Get(name)
+	if found {
+		cached := value.(*localCachedFSValue)
+		file, err = fs.backing.ReadFile(cached.name)
+		// If for some reason we can't read the backing store, make sure we remove from cache
+		if err != nil {
+			fs.cache.Del(name)
+		}
+	}
+	err = fmt.Errorf("%s unexpectedly missing", name)
+	return
+}
+
+func (fs *localCacheFSLayer) WriteFile(name string, content io.Reader) (written int64, err error) {
+	written, err = fs.backing.WriteFile(name, content)
+	if err != nil {
+		return
+	}
+
+	if !fs.cache.SetWithTTL(name, &localCachedFSValue{name: name, modtime: time.Now()}, written, fs.TTL) {
+		fs.remove(name)
```

Comment:
I really don't like having to do this, but I felt worse reading the data just to get the size, so I could set in cache, before I do a copy. It's all very weird, but if we had the size as input, I would feel much better, which to be honest, I think we do. All the consumers of this function are working with Buffer and []byte, so they all have a size. I could leave this method just to future proof the case of not having a size, will see.

Reply:
Ok, latest change adds WriteData method which is complementary to WriteFile. Though since all this is specific to esm.sh, I'm close to just replacing WriteFile with the WriteData signature.

Reply:
`WriteData` makes sence if we storage esm.sh builds to shared space. the offical esm.sh server use a  NFS disk as the storage that can be extended dynamic and can be shared in multiple mechines

Reply:
I'm still mentally going a little back and forth on just removing / replacing the current `WriteFile` signature with `WriteData`, though I'm probably overthinking it.